kernel/src/fs/epoll/epoll_file.rs

```diff
@@ -215,6 +221,47 @@ impl EpollFile {
         }
     }
 
+    fn collect_ready(&self, max_events: usize, ep_events: &mut Vec<EpollEvent>) {
+        let pop_guard = self.pop_guard.lock();
+
+        let mut limit = None;
+
+        loop {
+            if ep_events.len() >= max_events {
+                break;
+            }
+
+            // Since we're holding `pop_guard`, no one else can pop the entries from the ready
+            // list. This guarantees that `pop_ready` will pop the ready entries we see when
+            // `collect_ready` starts executing, so that such entries are never duplicated.
```

Comment:
Sorry, I can't get the comment why entries will be duplicated. If without `pop_guard`, are the dupliacated entries added by line 260?

Reply:
Yes. Race conditions. If we remove `pop_guard` and two programs do `pop_ready` concurrently, there will be race conditions.

For example,
```
<Thread 1>                <Thread 2>

limit = None              limit = None

pop_ready()
[ Entry X, Y are ready ]
entry = X
limit = Some(2)

                          pop_ready()
			  [ Entry Y is ready ]
			  entry = Y
			  limit = Some(1)

------- Now Entry X becomes ready again ------

pop_ready()
[ Entry X is ready ]
entry = X
limit = Some(0)
```

As emphasized in the original comments, the important fact is: This guarantees that `pop_ready` will **pop the ready entries we see when `collect_ready` starts executing**. These entries are never duplicated.

Reply:
Thanks for your explanation!